### PR TITLE
No more leaking of mounttype via metadata or icon in public shares.

### PR DESCRIPTION
**Pull request by rullzer**
Patch for #13789

There are no unit tests. I could not find unit tests for the ajax/list.php stuf, which makes sense since it is far from trivial to test.

### CODE REVIEW

**Comment by LukasReschke** [comment]
I'd prefer if we could at this point also move to an AppFramework controller here, because then we're able to test this ;-)

See https://github.com/owncloud/core/tree/master/apps/files_sharing/lib/controllers and https://github.com/owncloud/core/blob/master/apps/files_sharing/appinfo/routes.php. Some easier examples how to use the AppFramework can be found at http://doc.owncloud.org/server/8.0/developer_manual/app/tutorial.html

Interested to adjust this as well? – If you need some hints or help just ask and you shall receive :smile:

**Comment by rullzer** [comment]
Sure, I will have a look

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+								OCP\ISession $session) {
+		parent::__construct($appName, $request);
+
+		$this->appConfig = $appConfig;
+		$this->session = $session;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: I'd remove that. It's always better to have the CSRF check executed even if not technically 100% needed as a defense-in-depth against yet unknown vectors.

If you use the jQuery AJAX functions ownCloud adds the adder automatically.

**Comment by rullzer** [comment]
I just added some commits to move to the appframework for the ajax/list.php
In order to do this the helper class needed to be non static.

I started with writing unit tests. Maybe more a required? So we can verify that everything keeps working!

I removed the mixing of GET and POST. Since as far as I could tell we were using the session to authenticate with passwords anyway. As least in my debug sessions I never came across a password begin posted.

Feedback is welcome!

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	 * @PublicPage
+	 *
+	 * @param string t
+	 * @param string dir
+	 * @param string sort
+	 * @param string sortdirection
+	 */
+	public function getList($t, $dir, $sort = 'name', $sortdirection = 'desc') {
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Please add on multiple lines and replace `$t` with something like `$token` to make it easier to understand :)

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ * @copyright 2015 Roeland Jago Douma
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: New header please, see https://github.com/owncloud/core/pull/14426

File: apps/files_sharing/tests/controller/ajaxcontroller.php [hunk 4]
```diff
@@ -0,0 +1,206 @@
+<?php
+/**
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ * @copyright 2015 Roeland Jago Douma
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
```
Comment (LukasReschke) [inline apps/files_sharing/tests/controller/ajaxcontroller.php]: New header please, see https://github.com/owncloud/core/pull/14426

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+		$data['files'] = $formattedFiles;
+		$data['dirToken'] = $linkItem['token'];
+
+		$permissions = $linkItem['permissions'];
+		if ($this->appConfig->getValue('core', 'shareapi_allow_public_upload', 'yes') === 'no') {
+			$permissions = \OCP\Constants::PERMISSION_READ;
+		}
+		$data['permissions'] = $permissions;
+
+		return new JSONResponse(
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Let's use a `DataResponse` instead which is more flexible as we could add more response formatters in the future.

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		} catch (\Exception $e) {
+			return new JSONResponse(array(), Http::STATUS_INTERNAL_SERVER_ERROR);
+		}
+
+		$linkItem = $data['linkItem'];
+		$absdir = $data['realPath'];
+		$absdir = Filesystem::normalizePath($absdir);
+
+		if (!Filesystem::is_dir($absdir . '/')) {
+			return new JSONRespone(
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Also DataResponse

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+		$helper = new Helper2($this->session);
+		try {
+			$data = $helper->setupFromToken($t, $dir);
+		} catch (InvalidTokenException $e) {
+			return new JSONResponse(array(), Http::STATUS_NOT_FOUND);
+		} catch (AuthenticationException $e) {
+			return new JSONResponse(array(), Http::STATUS_FORBIDDEN);
+		} catch (\Exception $e) {
+			return new JSONResponse(array(), Http::STATUS_INTERNAL_SERVER_ERROR);
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: DataResponse as stated in the other comments

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 * @PublicPage
+	 *
+	 * @param string t
+	 * @param string dir
+	 * @param string sort
+	 * @param string sortdirection
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: PHPDoc requires to use `$` in front of variable names

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+namespace OCA\Files_Sharing\Controllers;
+
+use OC;
+use OCP;
+use OCP\AppFramework\Controller;
+use OCP\IRequest;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\JSONResponse;
+use OCA\Files_Sharing\Helper2;
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Not existing file? – Missed to commit? – Might also make sense to rename this into some better name :)

File: apps/files_sharing/lib/controllers/ajaxcontroller.php (hunk 6 above)
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: It's also a typo in here. Which IDE do you use? PHPStorm should show such errors directly to you :)

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		$files = \OCA\Files\Helper::getFiles($absdir, $sortAttribute, $sortDirection);
+
+		$formattedFiles = array();
+		foreach ($files as $file) {
+			$entry = \OCA\Files\Helper::formatFileInfo($file);
+			// for now
+			unset($entry['directory']);
+			// do not disclose share owner
+			unset($entry['shareOwner']);
+			// do not disclose if somethign is a remote shares
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: `something`

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+		} catch (InvalidTokenException $e) {
+			return new JSONResponse(array(), Http::STATUS_NOT_FOUND);
+		} catch (AuthenticationException $e) {
+			return new JSONResponse(array(), Http::STATUS_FORBIDDEN);
+		} catch (\Exception $e) {
+			return new JSONResponse(array(), Http::STATUS_INTERNAL_SERVER_ERROR);
+		}
+
+		$linkItem = $data['linkItem'];
+		$absdir = $data['realPath'];
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: While at it rename to `$absDir` for better readability.

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * Class AjaxController
+ *
+ * @package OCA\Files_Sharing\Controllers
+ */
+class AjaxController extends Controller {
+
+	/** @var AppConfig */
+	protected $appConfig;
+	/** @var OCP\Session */
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: That is an `ISession`

File: apps/files_sharing/appinfo/routes.php [hunk 13]
```diff
@@ -30,7 +30,10 @@
 $application->registerRoutes($this, [
 	'resources' => [
 		'ExternalShares' => ['url' => '/api/externalShares'],
-	]
+	],
+	'routes' => array(
```
Comment (LukasReschke) [inline apps/files_sharing/appinfo/routes.php]: I'd prefer to stay consistent and use `[` (array syntax for PHP >= 5.4)

Comment (LukasReschke) [inline apps/files_sharing/appinfo/routes.php]: Same for other usages of `array` in the controller below.

File: apps/files_sharing/appinfo/app.php [hunk 14]
```diff
@@ -41,6 +41,8 @@
 
 // Exceptions
 OC::$CLASSPATH['OCA\Files_Sharing\Exceptions\BrokenPath'] = 'files_sharing/lib/exceptions.php';
+OC::$CLASSPATH['OCA\Files_Sharing\Exceptions\InvalidTokenException'] = 'files_sharing/lib/exceptions.php';
```
Comment (LukasReschke) [inline apps/files_sharing/appinfo/app.php]: :see_no_evil:

File: apps/files_sharing/application.php [hunk 15]
```diff
@@ -67,6 +68,14 @@ public function __construct(array $urlParams=array()){
 				$c->query('ExternalManager')
 			);
 		});
+		$container->registerService('AjaxController', function(SimpleContainer $c) use ($server) {
+			return new AjaxController(
+				$c->query('AppName'),
+				$c->query('Request'),
+				$server->getAppConfig(),
```
Comment (LukasReschke) [inline apps/files_sharing/application.php]: @DeepDiver1975 @MorrisJobke  What is the state here? – Should people use IConfig or IAppConfig or whatever?

**Comment by LukasReschke** [comment]
Thanks @rullzer. Great start! :smile:

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+ * @copyright 2015 Roeland Jago Douma
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later.
+ * See the COPYING-README file.
+ */
+
+namespace OCA\Files_Sharing\Controllers;
+
+use OC;
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Really required?

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+use OCP;
+use OCP\AppFramework\Controller;
+use OCP\IRequest;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\JSONResponse;
+use OCA\Files_Sharing\Helper2;
+use OC\Files\Filesystem;
+use OC\AppConfig;
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Private API :see_no_evil:  - please try to use only functions in `OCP` if possible :)

File: apps/files_sharing/application.php [hunk 18]
```diff
@@ -67,6 +68,14 @@ public function __construct(array $urlParams=array()){
 				$c->query('ExternalManager')
 			);
 		});
+		$container->registerService('AjaxController', function(SimpleContainer $c) use ($server) {
```
Comment (LukasReschke) [inline apps/files_sharing/application.php]: I'd go with another not so generic name to make more clear what this controller is for. Maybe `FileListController` or something like that…?

File: apps/files_sharing/tests/controller/ajaxcontroller.php [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	/** @var IAppContainer */
+	private $container;
+	/** @var string */
+	private $user;
+	/** @var string */
+	private $token;
+	/** @var string */
+	private $oldUser;
+	/** @var string */
+	private $oldSERVER;
```
Comment (LukasReschke) [inline apps/files_sharing/tests/controller/ajaxcontroller.php]: PHP is case-sensitive when it comes to variables, below you used `oldServer` and here it is `oldSERVER`

(though the code might have worked before as well because PHP would magically have created it :see_no_evil:)

File: apps/files_sharing/tests/controller/ajaxcontroller.php [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+
+	public function testInvalidShareToken() {
+		$response = $this->ajaxController->getList($this->token.'x', '/');
+
+		$this->assertCount(0, $response->getData());
+		$this->assertEquals(404, $response->getStatus());
+	}
+
+	public function testInvalidAuth() {
+		$linkItem = \OCP\Share::getShareByToken($this->token, false);
```
Comment (LukasReschke) [inline apps/files_sharing/tests/controller/ajaxcontroller.php]: `$linkItem` is unused?

**Comment by rullzer** [comment]
@LukasReschke thanks for the feedback. I already made a bunch of commits to fix some of them. The others I'll have a look at later.

Commits will be nicely merged eventually of course :)

File: apps/files_sharing/application.php (hunk 18 above)
Comment (rullzer) [inline apps/files_sharing/application.php]: I'm fine with that. However I named it Ajax controller to eventually move all the stuff from the ajax dir to a testable controller. And then a more generic name seems better.

I'm fine either way.

File: apps/files_sharing/appinfo/app.php (hunk 14 above)
Comment (rullzer) [inline apps/files_sharing/appinfo/app.php]: It seems this line is only required to keep phpunit happy... So I probably need to add it somewhere else s phpunit is happy.

File: apps/files_sharing/lib/helper2.php [hunk 21]
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: Author and Co. missing

File: apps/files_sharing/lib/helper2.php [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+	 * @param string $password optional password
+	 *
+	 * @throws InvalidTokenException
+	 */
+	public function setupFromToken($token, $relativePath = null, $password = null) {
+		\OC_User::setIncognitoMode(true);
+
+		$linkItem = \OCP\Share::getShareByToken($token, is_null($password) ? false : true);
+		if($linkItem === false || ($linkItem['item_type'] !== 'file' && $linkItem['item_type'] !== 'folder')) {
+			\OC_Log::write('core-preview', 'Passed token parameter is not valid', \OC_Log::DEBUG);
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: Can use ILogger which could get injected

File: apps/files_sharing/lib/helper2.php [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+		if(!isset($linkItem['uid_owner']) || !isset($linkItem['file_source'])) {
+			\OC_Log::write('core-preview', 'Passed token seems to be valid, but it does not contain all necessary information . ("' . $token . '")', \OC_Log::WARN);
+			throw new IncompleteTokenException();
+		}
+
+		$rootLinkItem = \OCP\Share::resolveReShare($linkItem);
+		$path = null;
+		if (isset($rootLinkItem['uid_owner'])) {
+			\OCP\JSON::checkUserExists($rootLinkItem['uid_owner']);
+			\OC_Util::tearDownFS();
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: JSON functions are deprecated

File: apps/files_sharing/lib/helper2.php [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+			\OC_Util::setupFS($rootLinkItem['uid_owner']);
+			$path = \OC\Files\Filesystem::getPath($linkItem['file_source']);
+		}
+
+		if ($path === null) {
+			\OCP\Util::writeLog('share', 'could not resolve linkItem', \OCP\Util::DEBUG);
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: ILogger

File: apps/files_sharing/lib/helper2.php [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+			throw new LinkItemException();
+		}
+
+		if (!isset($linkItem['item_type'])) {
+			\OCP\Util::writeLog('share', 'No item type set for share id: ' . $linkItem['id'], \OCP\Util::ERROR);
+			throw new ItemTypeException();
+		}
+
+		if (isset($linkItem['share_with']) && (int)$linkItem['share_type'] === \OCP\Share::SHARE_TYPE_LINK) {
+			if (!self::authenticate($linkItem, $password)) {
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: $this?

File: apps/files_sharing/lib/helper2.php [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+		$this->session = $session;
+	}
+
+	/**
+	 * Sets up the filesystem and user for public sharing
+	 * @param string $token string share token
+	 * @param string $relativePath optional path relative to the share
+	 * @param string $password optional password
+	 *
+	 * @throws InvalidTokenException
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: Can throw other exceptions as well

File: apps/files_sharing/lib/helper2.php [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+					 * @link https://github.com/owncloud/core/issues/10671
+					 */
+					if(!empty($newHash)) {
+
+					}
+				} else {
+					return false;
+				}
+			} else {
+				\OCP\Util::writeLog('share', 'Unknown share type '.$linkItem['share_type']
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: ILogger

File: apps/files_sharing/lib/helper2.php (hunk 22 above)
Comment (rullzer) [inline apps/files_sharing/lib/helper2.php]: Ah yes that was still on my todo list. Hopefully later tonight

**Comment by rullzer** [comment]
@LukasReschke thanks for the comments. i have to admit helper2.php is still mostly a copy of helper.php.. More updates will follow (hopefully) soon.

File: apps/files_sharing/lib/controllers/ajaxcontroller.php (hunk 2 above)
Comment (rullzer) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: I agree that woul be better. However all the getList stuff uses t=.... chaning this to token breaks routing... or has to be done in a lot of places at once (a lot of which are curerntly not unit tested)

Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: *sigh*

File: apps/files_sharing/lib/controllers/ajaxcontroller.php (hunk 17 above)
Comment (rullzer) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: This is a leftover from the original list.php
I did a quick search but could not find a public API for this. Did I miss anything or is it just not available?

Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Use IConfig->getAppValue etc…

Comment (rullzer) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Ah yes that one I know.
I ment for OC\Files\Filesystem::normalizePath

Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Ahh… Yes… Let me look if this is actually required here still or if those checks are duplicated… Later ;-)

Comment (rullzer) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: I noticed that in helper2 there are more uses of the OCP classes... but I think it is unavoidable there

**Comment by LukasReschke** [comment]
Rebase :)

File: apps/files_sharing/appinfo/routes.php [hunk 28]
```diff
@@ -30,6 +31,9 @@
 $application->registerRoutes($this, [
 	'resources' => [
 		'ExternalShares' => ['url' => '/api/externalShares'],
+	],
+	'routes' => [
+		['name' => 'ajax#get_list', 'url' => '/ajax/list.php', 'verb' => 'GET'],
```
Comment (LukasReschke) [inline apps/files_sharing/appinfo/routes.php]: Maybe time to get rid of that redundant `.php` as well and change it in all JS files? :see_no_evil:

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+
+	/**
+	 * @NoAdminRequired
+	 * @PublicPage
+	 *
+	 * @param string $t Token of the share
+	 * @param string $dir Subdirectory requested
+	 * @param string $sort Sort attribute
+	 * @param string $sortdirection Sort direction
+	 */
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: `@return DataResponse`

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+			$entry['permissions'] = \OCP\Constants::PERMISSION_READ;
+			$formattedFiles[] = $entry;
+		}
+
+		//$data['directory'] = $relativePath; <-- where does this come from?
+		$data['files'] = $formattedFiles;
+		$data['dirToken'] = $linkItem['token'];
+
+		$permissions = $linkItem['permissions'];
+		if ($this->appConfig->getValue('core', 'shareapi_allow_public_upload', 'yes') === 'no') {
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: Always use IConfig if possible please :)

File: apps/files_sharing/lib/controllers/ajaxcontroller.php [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+			// do not disclose share owner
+			unset($entry['shareOwner']);
+			// do not disclose if something is a remote shares
+			unset($entry['mountType']);
+			unset($entry['icon']);
+			$entry['permissions'] = \OCP\Constants::PERMISSION_READ;
+			$formattedFiles[] = $entry;
+		}
+
+		//$data['directory'] = $relativePath; <-- where does this come from?
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/ajaxcontroller.php]: https://github.com/owncloud/core/pull/14341/files#diff-0c38edccfcc1f9aa0e2d2c9a16734085L41

File: apps/files_sharing/lib/controllers/sharecontroller.php [hunk 32]
```diff
@@ -1,7 +1,7 @@
 <?php
 /**
  * @author Bjoern Schiessle <schiessle@owncloud.com>
- * @author Georg Ehrke <georg@owncloud.com>
+ * @author Georg Ehrke <georg@ownCloud.com>
```
Comment (LukasReschke) [inline apps/files_sharing/lib/controllers/sharecontroller.php]: Unrelated?

File: apps/files_sharing/lib/helper2.php [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+use OCA\Files_Sharing\Exceptions\InvalidUserException;
+
+class Helper2 {
+
+	/** @var ISession */
+	protected $session;
+	/** @var ILogger */
+	protected $logger;
+
+	public function __construct(ISession $session,
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: PHPDoc please.

File: apps/files_sharing/lib/helper2.php [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+	 * Sets up the filesystem and user for public sharing
+	 * @param string $token string share token
+	 * @param string $relativePath optional path relative to the share
+	 *
+	 * @throws InvalidTokenException
+	 * @throws IncompleteTokenException
+	 * @throws LinkItemException
+	 * @throws ItemTypeException
+	 * @throws AuthenticationException
+	 */
```
Comment (LukasReschke) [inline apps/files_sharing/lib/helper2.php]: screen shot 2015-02-28 at 21 40 40

File: apps/files_sharing/tests/controller/ajaxcontroller.php [hunk 35]
```diff
@@ -0,0 +1,199 @@
...
+use OCA\Files_Sharing\Application;
+use OCP\AppFramework\IAppContainer;
+use OCP\Security\ISecureRandom;
+use OC\Files\View;
+use OCP\Share;
+
+/**
+ * @package OCA\Files_Sharing\Controllers
+ */
+class AjaxControllerTest extends \PHPUnit_Framework_TestCase {
```
Comment (LukasReschke) [inline apps/files_sharing/tests/controller/ajaxcontroller.php]: New tests should extend `\Test\TestCase`

File: apps/files_sharing/tests/controller/ajaxcontroller.php [hunk 36]
```diff
@@ -0,0 +1,199 @@
...
+
+use OC\Files\Filesystem;
+use OCA\Files_Sharing\Application;
+use OCP\AppFramework\IAppContainer;
+use OCP\Security\ISecureRandom;
+use OC\Files\View;
+use OCP\Share;
+
+/**
+ * @package OCA\Files_Sharing\Controllers
```
Comment (LukasReschke) [inline apps/files_sharing/tests/controller/ajaxcontroller.php]: Namespace looks wrong?

**Comment by LukasReschke** [comment]
Looks already really good otherwise! – Thanks!!! :rocket: :crown:

File: apps/files_sharing/appinfo/routes.php (hunk 28 above)
Comment (rullzer) [inline apps/files_sharing/appinfo/routes.php]: Well, that would be nicer. But would require the other ajax calls to be rewritten as well. So once that is done it seems like a good idea.

**Comment by rullzer** [comment]
@PVince81 @MorrisJobke can I ask you both to take a look?

**Comment by MorrisJobke** [comment]
I just monkey tested this. I shared some files and folders (internal and public). Tried to access them. Looked at the logs. I couldn't find anything suspicious.

I'm fine with that, but I can't judge on the code.

:+1:

**Comment by PVince81** [comment]
Needs rebase

**Comment by MorrisJobke** [comment]
rebase needed :)

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
Jenkins is happy. See #15279 
Please review.

**Comment by MorrisJobke** [comment]
@LukasReschke @icewind1991 Review please :)

**Comment by LukasReschke** [comment]
I really appreciate and like the change but as we have freeze tomorrow and this change is very huge and touches magic sharing code I'd appreciate to re-evaluate this again this for 8.2.

Since the touched bits shouldn't change much (if at all) until then I think this should not result in a rebase horror trip :)

**Comment by rullzer** [comment]
Fine with me. It should be carefully reviewed since the rebase did create some conflicts. 

But we can merge it once we start working on 8.2.

**Comment by MorrisJobke** [comment]
@rullzer rebase needed

**Comment by PVince81** [comment]
@rullzer great work, but we'd like to get rid of ajax/*.php and have the web UI use WebDAV instead. See #12353 

Is there anything from this PR that we could still keep regardless ?

**Comment by PVince81** [comment]
Reminds me I need to find a way to return the mount type through PROPFIND...

**Comment by rullzer** [comment]
@PVince81 not really then. Or at least not something that requires major rebasing... I branched that original PR and will work on that when the todo list is smaller.

This PR is now tiny. But it does make sure the info is not leaked. I'm fine with not merging this as long as we keep in mind to do this properly for the webdav stuff :)

**Comment by PVince81** [comment]
Nice, I like its simplicity.

:+1: let's merge this for now

**Comment by MorrisJobke** [comment]
Tested and works :*1:

**Comment by MorrisJobke** [comment]
:+1:

**Comment by PVince81** [comment]
Here is the same for the public webdav endpoint: https://github.com/owncloud/core/pull/17605
